### PR TITLE
Gen2: Add spatial detection network and spatial location calculator

### PR DESCRIPTION
**Pull request by szabi-luxonis**
- Added spatial detection network node == DetectionNetwork + xyz spatial data.
- Added spatial location node == retrieves average depth + xyz spatial data on a set of predefined ROIs mapped to depth frame.

Added 3 examples:

- spatial_mobilenet - runs mobilenet ssd on RGB camera and retrieves XYZ data.
- spatial_mobilenet_mono - runs mobilenet ssd on rectified right camera (perfectly matches depth frame) and retrieves XYZ data.
- spatial_location_calculator - calculates average depth + XYZ on a predefined ROI, controllable w/ WASD keys.

- Fixed stereo example, rectified frame was flipped.

FW public changelog:

* Fixed logger in resource allocator, the log didn't reach the host on error.
* Fixed uninitialized values in ImageManip, the parent frame metadata wasn't copied (timestamp etc.)
* Added logger in stereo node, it reports back the configured settings and calibration info, together w/ homography

* Changed all stereo depth output format to U16
* Fixed issues in stereo node regarding to ImageFrames, instance number was not set properly.
* Undefined/invalid depth set to 0 instead of infinity (65535 mm)
* Fixed X,Y,Z coordinates, Z axis was wrongly decomposed on gen1.

Displayed depth frame now goes through MIN_MAX normalization + histogram equalization + colorization. Better for visualization.

The documentation will cover the API.

### CODE REVIEW

**Comment by Luxonis-Brandon** [comment]
AWESOME!  Very excited about this!

File: examples/src/spatial_location_calculator_example.cpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    auto spatialCalcConfigInQueue = d.getInputQueue("spatialCalcConfig");
+
+    cv::Mat depthFrame;
+    auto color = cv::Scalar(255, 255, 255);
+    std::cout << "Use WASD keys to move ROI!" << std::endl;
+
+    while(1) {
+        auto depth = depthQueue->get<dai::ImgFrame>();
+        depthFrame = cv::Mat(depth->getHeight(), depth->getWidth(), CV_16UC1, depth->getData().data());
+        cv::Mat depthFrameColor = cv::Mat(depth->getHeight(), depth->getWidth(), CV_8UC1);
```
Comment (themarpe) [inline examples/src/spatial_location_calculator_example.cpp:88]: Can these two cases be covered using `getFrame` API?
If not, can we add support?

File: examples/src/spatial_mobilenet_example.cpp [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    while(1) {
+        auto imgFrame = preview->get<dai::ImgFrame>();
+        auto det = detections->get<dai::SpatialImgDetections>();
+        auto depth = depthQueue->get<dai::ImgFrame>();
+
+        auto dets = det->detections;
+
+        cv::Mat depthFrame = cv::Mat(depth->getHeight(), depth->getWidth(), CV_16UC1, depth->getData().data());
+
+        cv::Mat depthFrameColor = cv::Mat(depth->getHeight(), depth->getWidth(), CV_8UC1);
```
Comment (themarpe) [inline examples/src/spatial_mobilenet_example.cpp:120]: As in previous example, use case for `getFrame` API?

File: examples/src/spatial_mobilenet_mono_example.cpp [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    spatialDetectionNetwork->setDepthUpperThreshold(5000);
+
+    spatialDetectionNetwork->setBlobPath(nnPath);
+
+    // Link plugins CAM -> NN -> XLINK
+    imageManip->out.link(spatialDetectionNetwork->input);
+    if(syncNN)
+        spatialDetectionNetwork->passthrough.link(xlinkOut->input);
+    else
+        imageManip->out.link(xlinkOut->input);
```
Comment (themarpe) [inline examples/src/spatial_mobilenet_mono_example.cpp:70]: Style: braces or inline

File: examples/src/spatial_mobilenet_mono_example.cpp [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+                auto ymax = (int)(roi.ymax * depth->getHeight());
+
+                cv::rectangle(depthFrameColor, cv::Rect(cv::Point(xmin, ymin), cv::Point(xmax, ymax)), color, cv::FONT_HERSHEY_SIMPLEX);
+            }
+        }
+
+        cv::Mat rectifiedRight = inRectifiedRight->getCvFrame();
+
+        if (flipRectified)
+            cv::flip(rectifiedRight, rectifiedRight, 1);
```
Comment (themarpe) [inline examples/src/spatial_mobilenet_mono_example.cpp:154]: Style: braces or inline

File: include/depthai/pipeline/datatype/SpatialLocationCalculatorConfig.hpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+   public:
+    SpatialLocationCalculatorConfig();
+    explicit SpatialLocationCalculatorConfig(std::shared_ptr<RawSpatialLocationCalculatorConfig> ptr);
+    virtual ~SpatialLocationCalculatorConfig() = default;
+
+    void setROIs(std::vector<SpatialLocationCalculatorConfigData> rois);
+    void addROI(SpatialLocationCalculatorConfigData& roi);
+
+    std::vector<SpatialLocationCalculatorConfigData> getConfigData() const;
+};
```
Comment (themarpe) [inline include/depthai/pipeline/datatype/SpatialLocationCalculatorConfig.hpp:25]: Document the public functions and class description

File: include/depthai/pipeline/datatype/SpatialLocationCalculatorData.hpp [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+   public:
+    SpatialLocationCalculatorData();
+    explicit SpatialLocationCalculatorData(std::shared_ptr<RawSpatialLocationCalculatorData> ptr);
+    virtual ~SpatialLocationCalculatorData() = default;
+
+    std::vector<SpatialLocationCalculatorDataOut> getDepthData() const;
+
+    std::vector<SpatialLocationCalculatorDataOut>& depth;
+};
```
Comment (themarpe) [inline include/depthai/pipeline/datatype/SpatialLocationCalculatorData.hpp:24]: Document the public functions and class description

File: include/depthai/pipeline/node/SpatialDetectionNetwork.hpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+     * Input message with data to be infered upon
+     * Default queue is blocking with size 5
+     */
+    using DetectionNetwork::input;
+
+    /**
+     * nput message with depth data used to retrieve spatial information about detected object
+     * Default queue is non-blocking with size 4
+     */
+    Input inputDepth{*this, "inputDepth", Input::Type::SReceiver, false, 4, {{DatatypeEnum::Buffer, true}}};
```
Comment (themarpe) [inline include/depthai/pipeline/node/SpatialDetectionNetwork.hpp:38]: `Input message with depthai data...` <- typo
Shouldn't this function take `ImgFrame` or does it intentionally takes `Buffer`?

File: examples/src/stereo_example.cpp [hunk 8]
```diff
@@ -130,11 +130,15 @@ int main(){
 
             if (outputRectified) {
                 auto rectifL = rectifLeftQueue->get<dai::ImgFrame>();
-                cv::imshow("rectified_left", cv::Mat(rectifL->getHeight(), rectifL->getWidth(),
-                        CV_8UC1, rectifL->getData().data()));
+                cv::Mat rectifiedLeftFrame = rectifL->getCvFrame();
```
Comment (themarpe) [inline examples/src/stereo_example.cpp:133]: You can try using `getFrame` here to do it without copying

File: examples/src/stereo_example.cpp [hunk 9]
```diff
@@ -130,11 +130,15 @@ int main(){
...
+                cv::flip(rectifiedLeftFrame, rectifiedLeftFrame, 1);
+                cv::imshow("rectified_left", rectifiedLeftFrame);
+
                 auto rectifR = rectifRightQueue->get<dai::ImgFrame>();
-                cv::imshow("rectified_right", cv::Mat(rectifR->getHeight(), rectifR->getWidth(),
-                        CV_8UC1, rectifR->getData().data()));
+                cv::Mat rectifiedRightFrame = rectifR->getCvFrame();
```
Comment (themarpe) [inline examples/src/stereo_example.cpp:138]: You can try using getFrame here to do it without copying

**Review by themarpe — CHANGES_REQUESTED** [review]

**Pull request merged**